sycl/include/syclcompat/group_utils.hpp

```diff
@@ -0,0 +1,1269 @@
+};
+
+template <typename U> struct base_traits<float, U> {
+  static constexpr U HIGH_BIT = U(1) << ((sizeof(U) * 8) - 1);
+  static __syclcompat_inline__ U twiddle_in(U key) {
+    U mask = (key & HIGH_BIT) ? U(-1) : HIGH_BIT;
+    return key ^ mask;
+  }
+  static __syclcompat_inline__ U twiddle_out(U key) {
+    U mask = (key & HIGH_BIT) ? HIGH_BIT : U(-1);
+    return key ^ mask;
+  }
+};
+
+template <typename T> struct traits : base_traits<T, T> {};
+template <> struct traits<uint32_t> : base_traits<uint32_t, uint32_t> {};
+template <> struct traits<int> : base_traits<int, uint32_t> {};
+template <> struct traits<float> : base_traits<float, uint32_t> {};
+
+template <int N> struct power_of_two {
+  enum { VALUE = ((N & (N - 1)) == 0) };
+};
+
+__syclcompat_inline__ uint32_t shr_add(uint32_t x, uint32_t shift,
+                                       uint32_t addend) {
+  return (x >> shift) + addend;
+}
+
+} // namespace detail
+
+/// Rearranging data partitioned across a work-group.
+///
+/// \tparam T The type of the data elements.
+/// \tparam ElementsPerWorkItem The number of data elements assigned to a
+/// work-item.
+template <typename T, size_t ElementsPerWorkItem> class exchange {
+public:
+  static size_t get_local_memory_size(size_t group_threads) {
+    size_t padding_values =
+        (INSERT_PADDING)
+            ? ((group_threads * ElementsPerWorkItem) >> LOG_LOCAL_MEMORY_BANKS)
+            : 0;
+    return (group_threads * ElementsPerWorkItem + padding_values) * sizeof(T);
+  }
+
+  exchange(uint8_t *local_memory) : _local_memory(local_memory) {}
+
+  // TODO: Investigate if padding is required for performance,
+  // and if specializations are required for specific target hardware.
+  static size_t adjust_by_padding(size_t offset) {
+
+    if constexpr (INSERT_PADDING) {
+      offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    }
+    return offset;
+  }
+
+  struct blocked_offset {
+    template <typename Item> size_t operator()(Item item, size_t i) {
+      size_t offset = item.get_local_linear_id() * ElementsPerWorkItem + i;
+      return adjust_by_padding(offset);
+    }
+  };
+
+  struct striped_offset {
+    template <typename Item> size_t operator()(Item item, size_t i) {
+      size_t offset = i * item.get_local_range(2) * item.get_local_range(1) *
+                          item.get_local_range(0) +
+                      item.get_local_linear_id();
+      return adjust_by_padding(offset);
+    }
+  };
+
+  template <typename Iterator> struct scatter_offset {
+    Iterator begin;
+    scatter_offset(const int (&ranks)[ElementsPerWorkItem]) {
+      begin = std::begin(ranks);
+    }
+    template <typename Item> size_t operator()(Item item, size_t i) const {
+      // iterator i is expected to be within bounds [0,VALUES_PER_THREAD)
+      return adjust_by_padding(begin[i]);
+    }
+  };
+
+  /// Inplace rearrange elements from blocked order to striped order.
```

Comment:
Can you please add documentation for all new functions you've added here https://github.com/intel/llvm/tree/sycl/sycl/doc/syclcompat, using the code comment documentation here or otherwise as a base (but following the style in the syclcompat README).

Reply:
Done.

Reply:
Thanks